### PR TITLE
Cleanup security voters cookbook recipes

### PR DESCRIPTION
**Pull request by wouterj**
| Q | A |
| --- | --- |
| Doc fix? | yes |
| New docs? | no |
| Applies to | all |
| Fixed tickets | #5084 |

Blacklisting using voters isn't a good practice and in fact, the article (`cookbook/security/voters`) was almost completely equal to the much newer voter article (`cookbook/security/voters_data_permission`). I've moved the only different piece (access decision strategy) to the newer voter article and then removed the old voter article.

As the old voter article had a much nicer URL than the new one, I then renamed the new article to have the better name and set up a redirection. A [quick search on Google](https://www.google.nl/search?q=%5Blink%3Ahttp%3A%2F%2Fsymfony.com%2Fdoc%2Fcurrent%2Fcookbook%2Fsecurity%2Fvoters.html%5D+-site%3Ahttp%3A%2F%2Fsymfony.com) revealed that there are almost no links to the old article talking about blacklisting specifically, just about voters. So I think we're good here.

The diff is horrible, that's why I've tried to create multiple commits (this will ease reviewing process). The only commit requiring review is 9c169c7 (as the other 2 are just moving and removing stuff).

### CODE REVIEW

File: cookbook/security/acl.rst [hunk 1]
```diff
@@ -14,7 +14,7 @@ the ACL system comes in.
     Using ACL's isn't trivial, and for simpler use cases, it may be overkill.
     If your permission logic could be described by just writing some code (e.g.
     to check if a Blog is owned by the current User), then consider using
-    :doc:`voters </cookbook/security/voters_data_permission>`. A voter is passed the object
+    :doc:`voters </cookbook/security/voters`. A voter is passed the object
```
Comment (xabbuh) [inline cookbook/security/acl.rst]: missing closing angle bracket

Comment (WouterJ) [inline cookbook/security/acl.rst]: fixed

File: book/security.rst [hunk 2]
```diff
@@ -929,10 +929,10 @@ other users. Also, as the admin user, you yourself want to be able to edit
 
 To accomplish this you have 2 options:
 
-* :doc:`Voters </cookbook/security/voters_data_permission>` allow you to
-  use business logic (e.g. the user can edit this post because they were
-  the creator) to determine access. You'll probably want this option - it's
-  flexible enough to solve the above situation.
+* :doc:`Voters </cookbook/security/voters>` allow you to use business logic
```
Comment (javiereguiluz) [inline book/security.rst]: `[...] to use business logic [...]` looks very generic to me. I'd like something like: `... apply your own business logic ...`, `... define the custom logic applied to ...`, etc.

File: cookbook/security/voters.rst [hunk 3]
```diff
@@ -1,205 +1,269 @@
...
+
+How Symfony Uses Voters
+-----------------------
+
+In order to use voters, you have to understand how Symfony works with them.
+All voters are called each time you use the ``isGranted()`` method on Symfony's
+security context (i.e. the ``security.context`` service). Each one decides
+if the current user should have access to some resource.
+
+Ultimately, Symfony uses one of three different approaches on what to do
```
Comment (javiereguiluz) [inline cookbook/security/voters.rst]: I'd reword it to expand the explanation. Something like the following:

```
Ultimately, Symfony takes the responses from all voters and makes the final
decission (to allow or deny access to the resource) according to the strategy defined
in the application, which can be: affirmative, consensus or unanimous.
```

File: cookbook/security/voters.rst [hunk 4]
```diff
@@ -1,205 +1,269 @@
...
+    {
+        public function supportsAttribute($attribute);
+        public function supportsClass($class);
+        public function vote(TokenInterface $token, $object, array $attributes);
+    }
 
-.. include:: /cookbook/security/voter_interface.rst.inc
+The :method:`Symfony\\Component\\Security\\Core\\Authorization\\Voter\\VoterInterface::supportsAttribute`
+method is used to check if the voter supports the given user attribute (i.e:
+a role like ``ROLE_USER``, an ACL ``EDIT``, etc.).
```
Comment (javiereguiluz) [inline cookbook/security/voters.rst]: Could we remove the `ACL` mention to not complicate things?

File: cookbook/security/voters.rst [hunk 5]
```diff
@@ -1,205 +1,269 @@
...
 
-In this example, you'll check if the user's IP address matches against a list of
-blacklisted addresses and "something" will be the application. If the user's IP is blacklisted, you'll return
-``VoterInterface::ACCESS_DENIED``, otherwise you'll return
-``VoterInterface::ACCESS_ABSTAIN`` as this voter's purpose is only to deny
-access, not to grant access.
+The :method:`Symfony\\Component\\Security\\Core\\Authorization\\Voter\\VoterInterface::supportsClass`
+method is used to check if the voter supports the class of the object whose
```
Comment (javiereguiluz) [inline cookbook/security/voters.rst]: `is used to check if [...]` -> `checks whether [...]`

File: cookbook/security/voters.rst [hunk 6]
```diff
@@ -1,205 +1,269 @@
...
 
-Creating a custom Voter
------------------------
+The :method:`Symfony\\Component\\Security\\Core\\Authorization\\Voter\\VoterInterface::vote`
+method must implement the business logic that verifies whether or not the
+user has access. This method must return one of the following values:
 
-To blacklist a user based on its IP, you can use the ``request`` service
-and compare the IP address against a set of blacklisted IP addresses:
+* ``VoterInterface::ACCESS_GRANTED``: The authorization will be granted by this voter;
```
Comment (javiereguiluz) [inline cookbook/security/voters.rst]: `will be granted` -> `is granted` ?

File: cookbook/security/voters.rst (hunk 4 above)
Comment (WouterJ) [inline cookbook/security/voters.rst]: Complicate things -> having 2 items in a list of examples? Not really sure how this can complicate anything, it just explains some example attributes.

File: cookbook/security/voters.rst [hunk 7]
```diff
@@ -1,205 +1,269 @@
...
-            $this->blacklistedIp = $blacklistedIp;
-        }
+        const VIEW = 'view';
+        const EDIT = 'edit';
 
         public function supportsAttribute($attribute)
         {
-            // you won't check against a user attribute, so return true
-            return true;
+            return in_array($attribute, array(
```
Comment (javiereguiluz) [inline cookbook/security/voters.rst]: Can we refactor this 4-lines array into 1-line array?

File: cookbook/security/voters.rst [hunk 8]
```diff
@@ -1,205 +1,269 @@
...
+            // get a Post instance
+            $post = ...;
 
-In this case, choose the ``unanimous`` strategy. Unlike the ``affirmative``
-strategy (the default), with the ``unanimous`` strategy, if only one voter
-denies access (e.g. the ``ClientIpVoter``), access is not granted to the
-end user.
+            // keep in mind, this will call all registered security voters
+            if (false === $this->get('security.context')->isGranted('view', $post)) {
+                throw new AccessDeniedException('Unauthorised access!');
```
Comment (javiereguiluz) [inline cookbook/security/voters.rst]: It should be `unauthorized` (American English)

File: cookbook/security/voters.rst (hunk 4 above)
Comment (javiereguiluz) [inline cookbook/security/voters.rst]: I mean removing any mention to the `ACL` concept. We're explaining voters here, which we always promote as the common sense alternative to the over-complicated ACLs. And by the way, as you know, ACLs will disappear soon from Symfony.

Comment (WouterJ) [inline cookbook/security/voters.rst]: *from Symfony core (I wouldn't describe doctrine, swiftmailer and monolog as removed from Symfony, while they are not included in core).

This paragraph is describing what `supportsAttribute` does. For this, one need to know what an attribute is. Mentioning only `ROLE_ADMIN` could add the confusion again that attributes are roles, which they certainly aren't.

File: book/security.rst (hunk 2 above)
Comment (dupuchba) [inline book/security.rst]: :+1:

**Comment by weaverryan** [comment]
@WouterJ Really good idea moving things around, etc - I like it a lot. I've merged this in and all the way up with minor tweaks at sha: d4afd3a9cbdace6ce6d1bab5bec1e66768f8c5b7 thanks to some suggestions.

Thanks!
